### ISSUE TITLE
Computationally stable softmax cost

### ISSUE THREAD

**Issue by bartvm** [bartvm/blocks#150]
Instead of calculating the cross-entropy on the softmax directly we should have a brick that calculates the cross-entropy with in a more numerically more stable way (as [in Pylearn2](https://github.com/lisa-lab/pylearn2/blob/master/pylearn2/models/mlp.py#L1418-L1419)). It makes a big difference (my current model gets `NaN` all over the place if I don't use it).

**Comment by rizar** [comment]
By the way, shouldn't that change be done in Theano? We are simply using Theano ops for softmax-style things.

**Comment by bartvm** [comment]
I guess that ideally it should be done, but my guess is that for some reason it's a hard optimization to implement, otherwise I'm sure they would have done it. It's quite essential though; models can really become untrainable without it, so I think we should make sure this is available quickly nonetheless.

**Comment by bartvm** [comment]
Currently being addressed in #326.

**Comment by bartvm** [comment]
Fixed in #326.
